File: py/Apple/Arrays_and_Strings/minWin.py
# S = "ADOBECODEBANC", T = "ABC"
from collections import Counter
def minWin(s: str,t: str):
    # using counter and missing 
    # to check whether t already inside s[start:end]
    t_counter = Counter(t)
    missing = len(t)
    i,j=0,0
    start,end = 0,0
    while (end < len(s)):
        if t_counter[s[end]]>0:
            missing-=1
        t_counter[s[end]]-=1
        
        end+=1

        if missing == 0:
            # find start
            while t_counter[s[start]]<0:
                t_counter[s[start]]+=1
                start +=1
            t_counter[s[start]]+=1
            missing+=1
            if j==0 or end - start < j - i:
                j,i=end,start
            start +=1
    return s[i:j]

if __name__ == "__main__":
    s = "ADOBECODEBANC"
    t = "ABC"
    print (minWin(s,t))
            

    

File: java/Apple/Arrays_and_Strings/minWin.java
package Apple.Arrays_and_Strings;
import java.util.Map;
import java.util.HashMap;


public class minWin {

    public String MinWin(String s, String t){
        Map<Character,Integer> map = new HashMap<Character,Integer>();
        int missing = t.length();
        // init map 
        for (int i = 0;i<t.length();++i){
            //int count = map.getOrDefault(s.charAt(i), 0);
            map.put(t.charAt(i),map.getOrDefault(t.charAt(i), 0)+1);
        }

        int start = 0, end = 0;
        int i=0, j=0;

        while (end < s.length()){
            if (map.getOrDefault(s.charAt(end), 0)>0) --missing;
            //int count = map.getOrDefault(s.charAt(end), 0);
            map.put(s.charAt(end),map.getOrDefault(s.charAt(end), 0)-1);

            ++end;

            if (missing==0){
                while(map.get(s.charAt(start))<0) {
                    map.put(s.charAt(start),map.get(s.charAt(start))+1);
                    ++start;
                }
                map.put(s.charAt(start),map.get(s.charAt(start))+1);
                ++missing;

                if (j==0 || end - start < j - i){
                    j=end;
                    i=start;
                }
                ++start;
            }
        }
        return s.substring(i,j);


    }



    public static void main(String[] args){
        String s = "ADOBECODEBANC", t = "ABC";
        minWin sol = new minWin();
        System.out.println(sol.MinWin(s, t));
    }
}